### PR TITLE
Features Allow binding async method to command

### LINKED ISSUE #2417: Binding Command to method can execute private methods
The private method will be executed.
```C#
private async Task OnOpen()
```
```C#
public async void OnOpen(string path)
```
```XAML
            <MenuItem Header="_Recent" Items="{Binding RecentProjects}">
                <MenuItem.Styles>
                    <Style Selector="MenuItem">
                        <Setter Property="Header" Value="{Binding Name}"/>
                        <Setter Property="Command" Value="{Binding $parent[UserControl].DataContext.Platform.OnOpen}"/>
                        <Setter Property="CommandParameter" Value="{Binding Path}"/>
                    </Style>
                </MenuItem.Styles>
            </MenuItem>
```

### PR DESCRIPTION
**Pull request by workgroupengineering**
## What does the pull request do?

Enable binding of commands to asynchronous methods with the following signatures

### nothing parameters
```csharp
async Task Do()
{
   await ...
}
```
```csharp
async ValueTask Do()
{
   await ...
}
```

### one object parameter

```csharp 
async Task Do(object parameter)
{
   await ...
}
```
```csharp 
async ValueTask Do(object parameter)
{
   await ...
}
```

### one typed parameter

```csharp 
async Task Do(string parameter)
{
   await ...
}
```
```csharp 
async ValueTask Do(string parameter)
{
   await ...
}
```

## What is the current behavior?

 do nothing

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Fixed issues
Fixes #2417

### CODE REVIEW

File: src/Avalonia.Base/Data/Converters/MethodToCommandConverter.cs [hunk 1]
```diff
@@ -28,14 +30,29 @@ public MethodToCommandConverter(Delegate action)
             var canExecuteMethodName = "Can" + action.Method.Name;
             var parameters = action.Method.GetParameters();
             var parameterInfo = parameters.Length == 0 ? null : parameters[0].ParameterType;
+            var isAsync = action.Method.ReturnType == typeof(Task);
```
Comment (maxkatz6) [inline src/Avalonia.Base/Data/Converters/MethodToCommandConverter.cs:33]: Should also handle ValueTask or any type with proper GetAwaiter

**Comment by grokys** [comment]
Not had chance to try this yet, but what happens when you bind a button to method that returns a long-running task and click the button again? Does it fire off two tasks? Or does the button get disabled until the first task has finished?

**Comment by workgroupengineering** [comment]
> 
> 
> Not had chance to try this yet, but what happens when you bind a button to method that returns a long-running task and click the button again? Does it fire off two tasks? Or does the button get disabled until the first task has finished?

I left the developer free to decide whether or not to disable the button based on the CanExecute method. See [here](https://github.com/AvaloniaUI/Avalonia/blob/8e1ed7bec3d9bd10801e8d6ffc9a8be1238a9156/samples/BindingDemo/ViewModels/DownloadViewModel.cs#L53-L87).

**Comment by workgroupengineering** [comment]
@grokys Can you review?

**Comment by grokys** [comment]
@workgroupengineering we're currently working on getting the last bugs ironed out for a 0.10 release, so we don't have much time, and new features are unlikely to go in until the new release it out. If you could hold on a little while, hopefully we can get the release out and review your PRs. Thanks for being patient!

**Comment by workgroupengineering** [comment]
Are there any chances that this PR will be merged?
I need it in my App and would like to use only official versions of Avalonia.

**Comment by grokys** [comment]
@workgroupengineering not sure yet, because as I say we're trying to get 0.10 finished off. However I'm not 100% sure we want to accept this, as it feels like it could become a foot-gun:

1. To prevent the async method being called multiple times as you say the user would have to implement the `CanExecute` method as well as the execute method
2. And so at this point (implementing two methods), this feels more complex to me than just using a `ReactiveCommand` which automatically disables the command while a request is running

Without any background as to why just using `ReactiveCommand` isn't sufficient I'm tempted to 👎 this. Not sure what others think, @MarchingCube @jkoritzinsky I suspect you might have some input here?

I guess it would help if you could give some background as to why you prefer this solution to using `ReactiveCommand`?

**Comment by workgroupengineering** [comment]
Thanks for the reply,

> @workgroupengineering not sure yet, because as I say we're trying to get 0.10 finished off.

I understand your reasons and I don't want to insist.

> * To prevent the async method being called multiple times as you say the user would have to implement the `CanExecute` method as well as the execute method

if you want I can reverse default the behavior and add an attribute to the execution method that allows you to execute it multiple times or extend AppBuilder to set it globally.

> 
> * And so at this point (implementing two methods), this feels more complex to me than just using a `ReactiveCommand` which automatically disables the command while a request is running

To use ReactiveCommand you must also expose a field or a property of type ICommand in the ViewModel and initialize them either in the ViewModel's switch or in a late way in the property.
With this method you have to have at most two methods and possibly a property that shows that there is processing in progress. 

Example

https://github.com/AvaloniaUI/Avalonia/blob/d095d36b7826dd0ac3e394ab85339cf9db3460bd/samples/BindingDemo/ViewModels/DownloadViewModel.cs#L1-L89





> I guess it would help if you could give some background as to why you prefer this solution to using `ReactiveCommand`?

I prefer not to use ReactiveCommand or similar
1 I don't like to implement properties in ViewModel just to export the command, making the ViewModel dirty
2 in general, I like to use as few dependencies on external libraries as possible, because if an external library contains a bug, you risk being blocked until the fix is ​​posted.

**Comment by jkoritzinsky** [comment]
I think this feature counts in my head as "too much magic". The command-to-method binding is meant to be a convenience. It's not meant to replace all usages of Command properties. The support you added to create a CanExecute method with reflection I was on the fence for because of the same reasons. In this case we're checking more even more state and holding some in the command.

In the project I'm working on at work, we're specifically working on removing "magic" in our new work and making everything explicit and controllable by the user. Complex stateful features like this make it really hard for new users to understand what's going on, and hard to diagnose issues.

**Comment by workgroupengineering** [comment]
Apple made it fortune with "it magic".

kidding aside, I agree that the developer needs to be in control of what happens, that's why the multiple call control of the can execute method needs to be controlled via CanExecute.

The MVVM methodology is giving way to other patterns precisely because it is verbose.

This PR aims to reduce it without the developer losing control.

In an experimental branch where I am studying CompiledBinding, I am trying to remove the "magic" by expnading the Execute and CanExecute methods in a class that derives from ICommand

**Comment by workgroupengineering** [comment]
At present 2e4c911 there are 2 NU5128 warnings in macOs, this depends on miss net461 build target for ControlCatalog.Desktop. Can I remove ControlCatalog.Desktop from non-Windows builds?

**Comment by grokys** [comment]
> The command-to-method binding is meant to be a convenience. It's not meant to replace all usages of Command properties. The support you added to create a CanExecute method with reflection I was on the fence for because of the same reasons.

Yeah agree with this 100%.

> The MVVM methodology is giving way to other patterns precisely because it is verbose.

That may be true, maybe we should get @JaggerJo's input here as he's working on [FuncUI](https://github.com/AvaloniaCommunity/Avalonia.FuncUI) which uses an MVU-based architecture. @JaggerJo what do you think? Would this help FuncUI?

**Comment by JaggerJo** [comment]
> > The command-to-method binding is meant to be a convenience. It's not meant to replace all usages of Command properties. The support you added to create a CanExecute method with reflection I was on the fence for because of the same reasons.
> 
> Yeah agree with this 100%.
> 
> > The MVVM methodology is giving way to other patterns precisely because it is verbose.
> 
> That may be true, maybe we should get @JaggerJo's input here as he's working on [FuncUI](https://github.com/AvaloniaCommunity/Avalonia.FuncUI) which uses an MVU-based architecture. @JaggerJo what do you think? Would this help FuncUI?

In FuncUI we rarely use binding's and commands. Everything is just F# code.

```fsharp
view (state: State) (dispatch: Msg) : IView =
    Button.create [
        Button.content "start long running process"
        Button.isDisabled state.IsBusy
        Button.onClick (fun args -> 
             // call whatever you want
        )
]
```

Looking at this from a C# Avalonia user I'm against adding it as it adds unnecessary complexity.

**Comment by workgroupengineering** [comment]
I disagree, because when the code is more readable with no unnecessary parts serving only as a link to the infrastructure, there is less chance of making mistakes. I follow the instructions of the team and close the PR.

**Pull request closed without merging**